[4-9단계 - 방탈출 예약 관리] 밍트(김명지) 미션 제출합니다.

Starlight258 opened the pull request:
안녕하세요 아서 😊 밍트입니다!

이번 미션도 잘 부탁드립니다! 🔥🔥
감사합니다 🙇‍♀️
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  [[BE] 레벨2 - 방탈출 예약 관리](https://prolog.techcourse.co.kr/studylogs/4228)

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [ ] 4
- [x] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->
데이터를 접근하는 영속화 계층을 추상화 수준에 따라 2개의 계층으로 나누었습니다.
sql을 직접 작성하여 실행하는 저수준 계층은 Dao에, dao를 조합하여 도메인 컬렉션을 다루는 계층은 Repository에 두었습니다.

## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [ ] 아니오. 전혀 경험이 없습니다.
- [ ] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [ ] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [x] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 어떤 부분에 집중하여 리뷰해야 할까요?
이번 미션에서는 데이터베이스를 적용하고, 레이어드 아키텍처를 이용하여 계층을 분리해보았는데, 잘 되었는지 궁금합니다!

<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

Starlight258 committed:
feat: 동시성 고려하여 ConcurrentHashMap 사용

Starlight258 committed:
feat: 반환타입 ResponseEntity로 감싸도록 수정

- Controller에서 객체를 바로 반환하는 것보다, ResponseEntity로 감싸 상태와 헤더 등을 설정할 수 있도록 한다.

Starlight258 committed:
feat: h2 데이터베이스 환경설정

Starlight258 committed:
feat: 예약 목록 조회 기능 구현

Starlight258 committed:
feat: 예약 추가/삭제 기능 구현

Starlight258 committed:
feat: 필요하지 않은 빈 주입 제거

- SimpleJdbcTemplate은 JdbcTemplate을 이용하여 생성할 수 있으므로 DataSource를 주입받을 필요가 없다.

Starlight258 committed:
feat: 시간 추가 기능 구현

Starlight258 committed:
feat: 시간 조회 기능 구현

Starlight258 committed:
feat: 예약 시간 삭제 기능 구현

Starlight258 committed:
docs: 시간 관리 기능 문서 작성

Starlight258 committed:
feat: 예약 기능 수정

- 시간을 시간 테이블에 저장된 값만 선택하도록 수정한다.

Starlight258 committed:
refactor: 비즈니스 로직 ReservationService로 분리

Starlight258 committed:
feat: 사용하지 않는 코드 제거

Starlight258 committed:
fix: Integer를 long으로 변환하는 문제 해결

- long으로 캐스팅하지 않고, .longValue()를 이용하여 변환한다.

Starlight258 committed:
test: text fixture를 이용해 중복 코드 제거

Starlight258 committed:
test: repository 테스트 작성

Starlight258 committed:
refactor: 추상화 레벨이 낮은 sql은 dao로 분리

- repository는 dao를 이용하여 도메인 객체를 조립한다.

Starlight258 committed:
test: dao, repository 테스트 작성

Starlight258 committed:
style: 필요없는 코드 제거 및 포맷팅

Starlight258 commented on src/main/java/roomescape/repository/ReservationRepository.java:
```diff
@@ -0,0 +1,14 @@
+
+    List<Reservation> findAll();
+
+    Reservation insert(String name, LocalDate date, long timeId);
```
## Repository 파라미터 타입별 장단점
### 파라미터 여러개 나열 (현재 방식)
- 단점
  - 파라미터가 달라지면(수, 종류 등)` Repository`에 영향을 준다.
  - 요청마다 각 메서드를 만들어야한다. 

### DTO 사용
- 단순히 파라미터가 많아질 경우 묶어주는 파라미터 객체이다.
- 파라미터 조합이 많아질 경우 (name+date로 조회, name+timeId로 조회 등등..) 그에 따라 `DTO`가 많아지고, 메서드도 많아져 유지보수하기 어렵다.

### 도메인 객체 (`Reservation` 사용)
- 요청에 필요하지 않은 필드는 `null`로 두어 `Reservation` 객체를 생성하여 요청을 보낸다.
- 장점 ) 요청에 따라 새로운 메서드를 만들지 않아도 된다.
- 단점 )
  - `Null Pointer Exception`이 발생할 수 있으며, 만약 요청하는 필드가 `null`이 불가능하다면 도메인 객체를 생성하여 파라미터로 둘 수 없다.
  - 요청만을 위해 도메인 객체를 무분별하게 사용하는 느낌이 든다. 도메인 객체로서 역할을 하지 않고 데이터 저장 역할만 수행하기 때문이다.

`ReservationRepository의` 메서드 행위에 따라서도 어떤 파라미터가 적합할지 달라질 수 있다는 생각도 들었습니다.

### 행위에 따른 메서드 타입
- `insert` : 도메인 객체를 생성하므로 필수 파라미터 모두 존재 ⮕ 도메인 객체로 반환 
- `find` : 몇개의 식별자로 찾으므로 단순 파라미터 나열

하지만 현재 `insert` 메서드는 파라미터 나열 방식인데요, 
> "요청만을 위해 도메인 객체를 무분별하게 사용하는 느낌이 든다. 도메인 객체로서 역할을 하지 않고 데이터 저장 역할만 수행하기 때문이다." 

이 부분으로 인해 주저하게 되어 도메인 객체를 파라미터로 전달하지 않았습니다.

아서는 어떠한 의견을 가지는지 궁금해요! 도메인 객체를 데이터 전달용으로 사용하는 것이 좋다고 생각하시나요?

Hyunta replied:
insert할 때 도메인 객체로 저장하는게 꺼려지는 이유가 id 값이 null일 수 있어서 이런 고민을 해주신 것 같네요.

> 도메인 객체를 데이터 전달용으로 사용하는 것이 좋다고 생각하시나요?

Service에서 도메인 객체를 생성하는 이유를 단순히 데이터 전달용이라고 생각하시나요?
Reservaion에 대한 이름이 무조건 영문으로 작성되어야 한다는 규칙이 추가되었다고 생각해봅시다.
지금 구조에서 어디서 관리할 수 있게 되나요? DTO에서도 해당 검증을 추가해줘야하고 Reservation 객체에서도 관리해줘야겠네요.
그럼 DB에는 의도치않게 도메인 생성할 때 이뤄져야하는 검증이 이뤄지지 않은채로 데이터가 들어가는 경우가 발생할 수 있습니다.
그렇다면 DTO에서 모든 검증을 다뤄야할까요?

`implementation 'org.springframework.boot:spring-boot-starter-validation:3.4.4'`
해당 의존성을 제거하고 다시 진행해봅시다, 지금은 멋진 코드를 만들기 보다는 스프링 기반에서 어떻게 동작하는지를 알아보는 것이 더 중요하다고 생각이 됩니다.

Starlight258 replied:
현재 코드의 문제는 검증이 되지 않은 필드가 `DB`에 들어가고, 그 이후에 `Reservation` 객체가 생성되는 것입니다.
따라서 검증을 `DB`에 들어가기 전에 수행해야 하는데요, 아서가 이야기해주신 부분은 검증을 `DTO`에 두게 되면 검증 로직이 도메인 객체와 `DTO`로 분산되어 유지보수하기 어렵다는 점입니다.

검증 위치를 도메인 객체로 둘 경우와 `DTO`로 둘 경우를 나누어 생각해보았습니다.
## 검증 위치
### 1.`DTO`로 두는 경우
```java
    Reservation insert(ReservationDBRequest reservationDbRequest);
```

```java
public record ReservationDBRequest(Name name, ReserationDate date, long timeId) {
}
```
- 각 필드에 대한 검증을 `값 객체`(`Name`, `ReservationDate`,..)등을 통해 수행하고, 검증이 완료된 값만 `Repository` 필드에 둔다.
- `Reservation`에서 필드 여러개를 복합적으로 검증해야할 경우, `DTO`에 검증 코드를 둘 수 있다. 하지만 그 경우 `DTO`와 `도메인 객체` 모두에 검증 코드가 중복되어 유지보수가 힘들다 (`DRY` 원칙 위배)

### 2. 도메인 객체로 둘 경우
- 검증 코드가 도메인 객체에만 존재하므로, 응집도가 높으며 중복코드가 발생하지 않아 유지보수가 쉽다.
- 엔티티가 생명 주기동안 식별성을 보장하진 않지만, jpa에서도 준영속 상태에서 db 저장 후 영속 상태가 되고, `@GeneratedValue`에서도 id를 db에 저장 전까지 `null`로 두므로 괜찮을 것 같다.
- 저장 전에 id로 구별되어야하는 상황이 없으므로 id가 null이어서 `null pointer exception`이 발생하는 경우가 없다.

따라서 아서와 여러 크루, 제 생각을 조합한 결과 도메인 객체로 전달하는 것이 더 낫다라는 결정을 하게 되었습니다.

[refactor: repository 파라미터 도메인 객체 전달하도록 수정](https://github.com/woowacourse/spring-roomescape-admin/pull/340/commits/cf1ce23a19a14a8b0d49c8e80c3c49cb120122cb)

---
## Validation
### 커스텀 어노테이션 생성
커스텀 어노테이션인 `@NotNull`을 생성하여 검증을 수행했습니다. 
```java
@Target({ElementType.FIELD, ElementType.PARAMETER})
@Retention(RetentionPolicy.RUNTIME)
public @interface NotNull {

    String message() default "null이 될 수 없습니다.";
}
```
[feat: NotNull 메타 어노테이션 추가](https://github.com/woowacourse/spring-roomescape-admin/pull/340/commits/83b427c234bc0b8378d6efee9ad64094b7ae429c)

그런데 커스텀 어노테이션은 요청 객체 생성 후 `리플렉션`으로 검증을 합니다. 따라서 검증 전까지 일시적으로 잘못된 요청 객체가 생성될 수 있는 문제가 있습니다. (`Validation` 라이브러리도 이러한 형태로 구현된 것으로 알고 있습니다.)

### 생성자에서 검증
```java
public record ReservationRequest(

        LocalDate date,
        String name,
        Long timeId
) {
    public ReservationRequest {
        if (date == null) {
            throw new IllegalArgumentException("[ERROR] 날짜는 필수입니다.");
        }
        if (name == null) {
            throw new IllegalArgumentException("[ERROR] 이름은 필수입니다.");
        }
        if (timeId == null) {
            throw new IllegalArgumentException("[ERROR] 시간은 필수입니다.");
        }
    }
```
따라서 객체의 유효성을 보장하기 위해 생성자에서 예외를 발생시키고, `GlobalExceptionHandler`로 전역 예외를 처리하도록 수정했습니다.

[refactor: 커스텀 어노테이션 `@NotNull` 제거 후 생성자 내부에서 예외 발생하도록 수정 + GlobalExceptionHandler](https://github.com/woowacourse/spring-roomescape-admin/pull/340/commits/e2d86fb69b01fb95fb0d4ab7ff62f58e8aa79b4f)

Hyunta replied:
학습한 내용을 잘 정리해주셨네요 👍🏻

Starlight258 commented on src/main/java/roomescape/repository/ReservationRepositoryImpl.java:
```diff
@@ -0,0 +1,36 @@
+@Repository
+public class ReservationRepositoryImpl implements ReservationRepository {
+
+    private final ReservationDao reservationDao;
```
### 영속화 계층 분리 - Dao, Repository
데이터를 접근하는 영속화 계층을 추상화 수준에 따라 2개의 계층으로 나누었습니다.

- `Dao`는 저수준 계층으로, 직접 sql을 생성하고 쿼리를 실행합니다.
- `Repository`는 고수준 계층으로, 도메인 객체의 컬렉션(저장소)처럼 동작합니다. 또한 여러 dao(datasource)를 조합하여 도메인 객체를 조립할 수 있습니다.

계층을 추상화 레벨에 따라 분리해보았는데, 이 설계에 대해 아서의 의견이 궁금합니다!

[(reference) dao vs repository](https://www.baeldung.com/java-dao-vs-repository)

Hyunta replied:
개인적으로는 과하다고 생각합니다.
DAO와 Repository 둘 중에 하나만 사용해도 충분하지 않나요?

Starlight258 replied:
좋은 설계와 과한 설계의 차이점에 대해 공부해보는게 좋을 것 같아 학습 후 정리해보았습니다.

## 좋은 설계
### **생각했던 좋은 설계**
- 협력을 기준으로 책임을 적절히 분리(`SRP`)
- 같은 추상화 수준별로 분리(구체적인 것 vs 추상적인 것, `DIP`)

그런데 글들을 읽어보면서, 좋은 설계란 **확장될 요구사항을 고려하는 것이 중요하다**는 것을 알게 되었습니다.

### **추가 고려할 사항**
- **미래에 추가될 요구사항을 최대한 변경 범위가 적도록 수정할 수 있는 설계**
  - 너무 먼 미래를 고려하지 않아야한다. 미리 도입한 추상은 변경 비용을 높인다. 
  - 미래의 변경 가능성이 불확실하다면, **더 단순하고 이해하기 쉬운 설계**를 선택한다.

### 설계시 고려해야 할 요구사항
- `join`하는 쿼리가 많아질 것 같다.
- `db 종류`가 바뀔 가능성이 적다. (`mysql`, `mongodb`)

## `Dao`와 `Repository`를 분리할 경우
- 테이블별로 `dao`를 나눈다면, 연관된 두 테이블을` join`하여 가져올 때 어떤 `dao`에 둘지 모호하다.
- `db 종류`가 바뀔 경우 `dao`를 구분하는 것이 좋지만, 당장은 바뀔 가능성이 적다.

### repository만 둘 경우(`Dao` 제거)
- `repository는` 도메인 객체를 중심으로 생각하므로, 여러 테이블에 걸친 `join` 쿼리도 특정 도메인 객체와 관련된 데이터를 가져오는 하나의 책임으로 정의할 수 있다. 즉 책임 경계가 명확해진다.
- 불필요한 계층이 줄어든다.

### 설계 변경해야할 시점 (dao와 repository를 나누어야할 시점)
- `db 종류`가 변경될 경우 `repository`내에 `dao`를 도입하여 필요한 부분만 분리한다.


따라서 `dao`와 `repository`를 분리하기보다는, `repository` 하나로 두어 쿼리를 작성하는 것이 낫다는 결론을 얻게 되었습니다.

이번 기회에 좋은 설계란 단순히 객체지향적으로 협력을 기준으로 책임을 분리하는 것 뿐만 아니라. **요구사항에 유연하게 대응해야한다**는 인사이트를 얻게 되었습니다. 감사합니다!

Hyunta replied:
설계라는게 참 아이러니한 부분들이 많은 것 같아요.
확장에 대해서 고려하지 않는 것도 문제가 되고, 과하게 고려하는 것 또한 문제가 됩니다.
그래서 하나로 딱 정의하기가 어려운 것 같습니다.

작성해주신 것처럼 쉬운 코드를 짜는 것이 좋은 접근이라고 생각합니다.
확장에 대한 필요가 생겼을 때 빠르게 파악하고, 수정할 수 있으니까요.
좋은 고민을 많이 해주셨네요 👍🏻

Starlight258 commented on src/main/java/roomescape/repository/ReservationRepositoryImpl.java:
```diff
@@ -0,0 +1,36 @@
+
+    @Override
+    public Reservation insert(final String name, final LocalDate date, final long timeId) {
+        long reservationId = reservationDao.insertReservation(name, date, timeId);
+        ReservationTime reservationTime = reservationTimeDao.findById(timeId);
+        return new Reservation(reservationId, name, date, reservationTime);
```
### `join` 쿼리를 `dao`에 두어도 되는가?
`dao`는 데이터 베이스 중심으로 설계되어, 보통 단일 테이블의 `crud`를 수행합니다.

크루들 앞에서 `Layered Architecture` 강의를 했었는데, 한 크루가 **"`join`시 두 테이블이 모두 필요한데, 이 작업은 dao에서 할 수 있나요? 쿼리 최적화의 경우 어떻게 동작하나요?"** 라고 물어보았습니다. 그런데 해당 쿼리를 어디에 둘지 고민이 되어 선뜻 답을 하기 어려웠습니다.

`dao`에서` join`을 한다고 가정하면, `repository`와` dao`의 차이점이 모호해지는 것 같다는 고민도 듭니다. 

`jpa`에서는 단순히 `repository`에서 메서드를 호출하여 많은 고민을 하지 않았는데,` jdbc`를 이용하다보니 데이터 접근 계층에 대한 고민이 많아지는 것 같습니다.🤔

**여러 테이블이 필요한 쿼리를 `dao`에서 수행해도 괜찮을까요?**

Hyunta replied:
여러 테이블이 필요한 쿼리를 DAO에서 수행하면 안된다고 생각하시는 이유는 무엇인가요?

---

> jpa에서는 단순히 repository에서 메서드를 호출하여 많은 고민을 하지 않았는데, jdbc를 이용하다보니 데이터 접근 계층에 대한 고민이 많아지는 것 같습니다.🤔

제 생각에 지금 밍트는 본인에게 큰 도움이 되는 학습 시간을 가질 수 있는 기회가 주어진 것 같습니다.
Spring과 JPA를 이미 접해보셔서 관습적으로 짜던 코드들이 일부 남아있는 것 같아요.
왜 그렇게 했었는지를 한번씩 되돌아보면서 본질적인 고민들을 해봅시다.

Starlight258 replied:
> 여러 테이블이 필요한 쿼리를 `DAO`에서 수행하면 안된다고 생각하시는 이유는 무엇인가요?

[dao vs repository](https://www.baeldung.com/java-dao-vs-repository)에서 `DAO`가 테이블 중심으로 설계가 되어 있다고 언급되어서 `dao`가 단일 테이블을 이야기한다고 생각했습니다.
> The Data Access Object Pattern, aka [DAO Pattern](https://www.baeldung.com/java-dao-pattern), is an abstraction of data persistence and is considered closer to the underlying storage, which is often table-centric.

따라서 여러 테이블이 필요한 `join`과 같은 쿼리는 `dao`의 책임을 벗어난다고 생각했습니다.


그런데 생각해보면 테이블 사이에 연관관계가 존재하므로 **여러 테이블이 연관된 쿼리가 필수적으로 발생**될 것이라고 생각합니다. 특히 성능 최적화를 위해 `join`을 많이 사용합니다.
따라서 `dao`에서 도메인 객체를 반환하기 위해 연관된 테이블까지 함께 조회할 수 있어야합니다.

즉 제가 오해했던 "**테이블 중심적**"이라는 말은, `DAO`가 단일 테이블만 다뤄야한다는 규칙이 아닌, `DAO`가 테이블 구조에 영향을 받는다(`dao에 테이블의 crud 연산이 포함된다`)는 의미인 것 같습니다. 

## JDBC, JPA
### JDBC
- JDBC는 `저수준 API`로, 쿼리를 직접 작성하고 실행한다.
### JPA
- JPA는 `고수준 API`로, 객체와 데이터베이스 간의 매핑을 자동으로 처리한다.(`ORM`) 즉 개발자가 직접 SQL을 작성할 필요가 없다.

`JPA`를 사용하다보니, `JDBC`를 사용하면서도 고수준 API를 `Repository`로 작성하고, 저수준 API는 `Dao`로 나누고 싶었던 것 같습니다. 또한 `JPA`에서는 복잡한 연관관계를 편하게 로딩할 수 있지만, `JDBC`에서는 직접 쿼리를 작성해야하므로 어떻게 효율적으로 연관관계를 로딩할지 후에 공부할 것 같습니다.

Hyunta replied:
DB 쿼리를 어떻게 요청하는 것이 좋을지를 고민해보면 쉽게 결론이 내려질 것 같습니다.
우선 DB에 대해서 학습을 해보시고, 어떻게 쿼리를 요청하는 것이 DB 입장에서 비용이 적게 들면서 안정적으로 가져올 수 있을지를 고민해봐주세요.

Starlight258 commented on src/main/java/roomescape/service/ReservationService.java:
```diff
@@ -0,0 +1,29 @@
+        this.reservationRepository = reservationRepository;
+    }
+
+    public List<Reservation> findAll() {
+        return reservationRepository.findAll();
+    }
```
### ApplicationService 테스트
서비스는 단순히` repository`를 호출하는 흐름 조정의 역할만 하고, 비즈니스 로직을 수행하지 않는데 테스트하는 것이 좋을지 궁금합니다!

어떤 로직을 테스트하고, 어떤 로직은 테스트하지 않는 것이 좋을까요?

제 생각엔 비즈니스 로직이 존재하는 객체와 외부 의존성이 있는 객체는 테스트하고, 그 외는 통합 테스트로 충분히 검증이 된다고 생각합니다.

테스트를 한다면, `Repository`를 `Fake 객체`로 만들어 주입하여 테스트할 것 같습니다!

Hyunta replied:
> 제 생각엔 비즈니스 로직이 존재하는 객체와 외부 의존성이 있는 객체는 테스트하고, 그 외는 통합 테스트로 충분히 검증이 된다고 생각합니다.

사용되어지는 단어가 모호해서 명확하게 해보시면 답을 찾으실 것 같습니다.
비즈니스 로직이 존재하는 객체, 외부 의존성이 있는 객체 -> ReservationService 아닌가요?
그 외 -> 어떤 클래스가 그 외에 들어갈 수 있나요?
통합 테스트 -> 어떤걸 통합테스트라고 말하시나요?

Starlight258 replied:
- **비즈니스 로직**이 존재하는 객체 : `도메인 객체`
- **외부 의존성**이 있는 객체 : `Repository`
- 외부 의존성을 사용하는 객체 : `Service`
- 그 외 : `Controller`

사실 서비스가 단순히 객체의 메서드를 호출하기 때문에 테스트 범위에 들어갈지 고민이 들었습니다.
그런데 생각해보면 서비스에서는 여러 도메인 객체와 리포지토리를 조합하여 **서비스 요구사항을 구현**합니다. 따라서 각 컴포넌트가 올바른 순서로 호출되는지, 예외 상황이 잘 발생되는지 테스트하는 것이 중요합니다.

- 단위 테스트 -> 클래스나 메서드(가장 작은 단위)를 격리된 환경에서 검증, 작지만 빠르고 반복적 실행에 중점 ex) `JUnit`, `Mockito`
- 통합 테스트 -> 여러 컴포넌트간의 상호작용 테스트, 전반적 시스템 검증 ex) `Spring Test`
- E2E 테스트 -> 사용자 관점에서 전체 시스템 기능 검증 (모든 계층을 포함한 완전한 시스템 테스트) ex) `RestAssured`

[test: service 테스트 작성](https://github.com/woowacourse/spring-roomescape-admin/pull/340/commits/1351366ab3f754b1ef8188b0264674065f53087a)

Hyunta replied:
서비스에서 특별한 로직은 없지만 조율하는 역할을 하고 있기 때문에 이것만으로도 충분히 테스트 대상이 된다고 생각합니다.
단위테스트가 전부 작동한다고 항상 뭉쳤을 때 정상적으로 작동하는 것은 아닐 수도 있기 때문입니다.
저는 단순 메서드 호출이라하더라도 있는 편이 안정감을 더 주는 것 같아요.

Starlight258 commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -13,38 +13,43 @@
 
 @RestController
 @RequestMapping("/reservations")
 public class ReservationController {
```
### Controller 테스트
`Controller`는 단순히 요청을 하고 응답을 하는 객체인데, 테스트하는 것이 좋을까요?
테스트한다면 어떤 부분을 테스트하는 것이 좋을까요? 응답 코드도 테스트 범주에 포함되는지 궁금해요!

Starlight258 commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -13,38 +13,43 @@
 
 @RestController
 @RequestMapping("/reservations")
 public class ReservationController {
```
### Controller 테스트
`Controller`는 단순히 요청을 하고 응답을 하는 객체인데, 테스트하는 것이 좋을까요?
테스트한다면 어떤 부분을 테스트하는 것이 좋을까요? 응답 코드도 테스트 범주에 포함되는지 궁금해요!

Hyunta replied:
밍트가 생각하실 때는 테스트가 필요하다고 생각하시나요?

Starlight258 replied:
제 생각엔 `controller`는 현재 단순히 요청에 대한 응답 상태 코드를 반환하고, `dto`로 변환하여 서비스를 호출하는 역할만 수행하고 있습니다.
상태코드는 `e2e` 테스트로 모두 검증되고 있으므로 `controller`는 테스트할 로직이 없다고 생각합니다.

Hyunta replied:
그렇다면 안해도 괜찮을 것 같습니다.
저도 마찬가지로 E2E 테스트로 충분하다고 생각합니다.

Hyunta commented on src/main/java/roomescape/controller/ReservationTimeController.java:
```diff
@@ -0,0 +1,48 @@
+
+    @PostMapping
+    public ResponseEntity<ReservationTimeResponse> addReservationTime(
+            @Valid @RequestBody ReservationTimeRequest timeRequest) {
```
`@Valid` 가 없었다면 어떻게 DTO를 검증하셨을 것인가요?

Starlight258 replied:
생성자에서 검증할 것 같습니다!

이전 코멘트에서 언급해주셨던 것처럼, validation 라이브러리를 지우고 검증을 수행하도록 수정했습니다!

Hyunta commented on src/main/java/roomescape/controller/ReservationTimeController.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    @DeleteMapping("/{id}")
+    public ResponseEntity<Void> delete(final @PathVariable long id) {
+        reservationTimeService.delete(id);
+        return ResponseEntity.ok().build();
+    }
```
삭제 응답에 대한 응답코드 200 말고 다른 선택지는 없을까요?

Starlight258 replied:
요구사항에 모두 200 상태코드로 정해져있어서 수정하지 않았지만 `delete`의 경우 `no content (204) `가 적절할 것 같습니다!

### HTTP Method 별 상태 코드
- GET : 200 OK
- POST : 201 Created
- PUT: 200 
- PATCH : 200 
- DELETE : 204
> 반환하고 싶은 상황에 따라 상태코드의 종류는 다양합니다.

[feat: http 메서드별 상태 코드 변경](https://github.com/woowacourse/spring-roomescape-admin/pull/340/commits/b879d512f3c8923d3c372008f509129911bd0d6a)

Hyunta commented on src/main/java/roomescape/repository/ReservationDao.java:
```diff
@@ -0,0 +1,64 @@
+            LocalTime time = LocalTime.parse(resultSet.getString("time_value"));
+            ReservationTime reservationTime = new ReservationTime(timeId, time);
+            return new Reservation(id, name, date, reservationTime);
+        });
```
RowMapper를 분리해놓으면 어떤가요?

Starlight258 replied:
상수로 선언하는 것이 훨씬 가독성이 좋아질 것 같습니다! 감사합니다.

Hyunta commented on src/test/java/roomescape/repository/ReservationDaoImplTest.java:
```diff
@@ -0,0 +1,69 @@
+import roomescape.domain.Reservation;
+import roomescape.fixture.TextFixture;
+
+@JdbcTest
```
<img width="940" alt="image" src="https://github.com/user-attachments/assets/dcb543ca-16a5-4084-9dc8-dff1e86f011f" />

요구사항을 철저히 지켜주세요, 가장 기본적이면서 중요한 철칙입니다.

Starlight258 replied:
`db` 테스트를 하기 위해 데이터베이스 연결을 해야하므로 `assertJ`로 수행할 수 없다고 막연히 생각했던 것 같습니다.
그런데 수동으로 테스트용 인메모리 데이터베이스를 연결하는 방법이 있다는 것을 알게 되었습니다!
따라서 코드에 반영하였습니다. 안될거라고 막연하게 생각하는 습관을 버리는게 좋을 것 같습니다. 감사합니다 !

[test:`@JdbcTest` 없이 테스트하기 위해 db 연결 설정 작성](https://github.com/woowacourse/spring-roomescape-admin/pull/340/commits/79fedad32965eb5d7ef170db095ddd73ae2b1146)

Hyunta commented on src/main/java/roomescape/repository/ReservationDao.java:
```diff
@@ -0,0 +1,64 @@
+                on r.time_id = t.id
+                """;
+        List<Reservation> query = jdbcTemplate.query(sql, (resultSet, rowNumber) -> {
+            long id = resultSet.getInt("id");
```
reservation_id로 받아야하지 않나요?
그리고 타입도 getLong으로 받아야할 것 같은데요?

Starlight258 replied:
맞습니다! 반영하였습니다!!

Hyunta commented on src/test/java/roomescape/MissionStepTest.java:
```diff
@@ -1,57 +1,177 @@
 package roomescape;
 
```
<img width="282" alt="image" src="https://github.com/user-attachments/assets/2cf9ae5b-2059-4889-a597-a6bf4f5c0780" />

테스트 몇개가 깨지는 것 같은데 확인 부탁드립니다.

Starlight258 replied:
혹시 어느 부분에서 깨지는지 알 수 있을까요? 제출 전에 모두 통과하는 것을 확인하고 미션 pr을 날렸는데 
환경 차이로 인해 테스트가 깨졌을 수 있을 것 같습니다.
<img width="391" alt="image" src="https://github.com/user-attachments/assets/65af1131-b840-4cd7-9951-2a8d46f2e2dc" />

에러 메세지 공유해주시면 바로 수정하겠습니다!!

Hyunta replied:
이건 IntelliJ로 테스트를 돌렸을 때 역직렬화 하면서 문제가 생기는 것 같네요.
무시하셔도 괜찮을 것 같습니다.

Hyunta reviewed (changes requested):
안녕하세요 밍트,
4~9단계 진행하시느라 고생하셨습니다.
몇가지 커멘트 남겨드렸는데 확인 부탁드립니다.
궁금하시거나 모호하신 부분이 있으면 DM 주세요.

Starlight258 committed:
refactor: repository 파라미터 도메인 객체 전달하도록 수정

- 도메인 객체를 전달하여 db에 유효한 데이터만 저장할 수 있도록 한다.
- 도메인 객체의 id 값이 null이 들어갈 수 있도록 Wrapper 타입으로 변경한다.

Starlight258 committed:
feat: NotNull 메타 어노테이션 추가

- validation 라이브러리를 지우고 직접 NotNull 어노테이션 생성

Starlight258 committed:
refactor: 커스텀 어노테이션(@NotNull) 제거 후 생성자 내부에서 예외 발생하도록 수정 + GlobalExceptionHandler

- 커스텀 어노테이션은 요청 객체 생성 후 리플렉션으로 검증을 한다. 따라서 검증 전까지 일시적으로 잘못된 요청 객체가 생성될 수 있다.
- 생성자 내부에서 예외를 발생시키면, 유효하지 않은 상태의 객체가 생성되는 것을 방지할 수 있다.
- 예외 처리는 GlobalExceptionHandler에서 공통적으로 수행한다.

Starlight258 committed:
refactor: 불필요한 dao 계층 제거

- 현재 repository와 dao를 분리함으로써 얻는 이점은 추상화 수준의 분리 뿐이다.
- 하지만 성능을 위해 join 쿼리가 많이 작성되고, 이 때  dao를 분리함으로써 경계가 모호해지므로 dao를 제거하고 repository 내에 모두 작성한다.

Starlight258 committed:
refactor: 필요없는 코드 제거 및 상수 추출

Starlight258 committed:
refactor: 메서드명 수정

Starlight258 committed:
test: service 테스트 작성

Starlight258 committed:
feat: http 메서드별 상태 코드 변경

Starlight258 committed:
test: @JdbcTest 없이 테스트하기 위해 db 연결 설정 작성

Starlight258 committed:
fix: 삭제시 id가 존재하지 않을 경우 예외 발생하도록 수정

Starlight258 commented:
@Hyunta 아서 리뷰 감사드립니다!!
피드백 주신 내용 모두 반영하였습니다! 
(테스트가 깨지는 부분은 에러 메세지 공유해주시면 바로 수정하겠습니다!!)

덕분에 설계적 관점과 jdbc, jpa 차이, layered architecture에 대해 다시 공부할 수 있었습니다. 

혹시 추가적인 개선사항이 있다면 피드백 주시면 감사드리겠습니다! 
많이 배웠습니다! 감사합니다. 🙇‍♀️

Hyunta reviewed (approved):
안녕하세요 밍트,
학습하신 내용을 잘 정리해주셨네요.
추가로 남겨주신 커멘트에 답변 남겨놨습니다.
이번 단계는 여기서 머지하도록 하겠습니다.
추가로 궁금하시거나, 모호한 부분이 있으면 DM 주세요.
